xla/service/cpu/tests/onednn_convolution_test.cc

```diff
@@ -32,159 +33,221 @@ limitations under the License.
-    reshape.2 = f32[1,11,11,1] reshape(convolution.0)
-    tuple.0 = (f32[1,11,11,1]) tuple(reshape.2)
-    ROOT get-tuple-element.0 = f32[1,11,11,1] get-tuple-element(tuple.0), index=0
-  })";
+  void SetUp() override {
+    if (!IsSupportedType(dtype)) {
+      GTEST_SKIP() << "CPU does not support " << dtypeString;
+    }
+  }
 
-  EXPECT_TRUE(RunAndCompare(convolution_module_str, ErrorSpec{1e-4, 1e-4}));
-  MatchOptimizedHlo(convolution_module_str, conv_rewrite_str_);
-}
+  void SetWeightsPrepacked(bool value) { weights_prepacked = value; }
+
+  void SetUserScratchpad(bool value) { user_scratchpad = value; }
+
+  std::string GetOptimizationsString() {
+    return (user_scratchpad || weights_prepacked)
+               ? absl::Substitute(conv_rewrite_optimizations_str_,
+                                  weights_prepacked, user_scratchpad)
+               : "";
+  }
 
-TEST_F(ConvolutionTest, Simple3DTestBF16) {
-  if (!IsSupportedType(PrimitiveType::BF16)) {
-    GTEST_SKIP() << "CPU does not support BF16.";
+  std::string ConvStringWithOptimizations(
+      const std::vector<absl::string_view> fused_ops) {
+    std::ostringstream stream;
+    std::for_each(
+        fused_ops.begin(), fused_ops.end(),
+        [&](const absl::string_view& arg) { stream << "\"" << arg << "\","; });
+    std::string fusions = stream.str();
+    if (fused_ops.size() > 0) {
+      fusions.pop_back();
+      return absl::Substitute(
+          conv_rewrite_str_,
+          absl::Substitute(conv_rewrite_fusions_str_, fusions),
+          GetOptimizationsString());
+    }
+    return absl::Substitute(conv_rewrite_str_, "", GetOptimizationsString());
   }
 
-  const char* convolution_module_str = R"(
-  HloModule convolution.test.bf16
+  // TODO(intel-tf): Remove this and simplify patterns when Elemental BF16 is
+  // fully supported.
+  PrimitiveType PromotedDtype() {
+    // BF16 is promoted to F32 because not all HLO Instructions currently
+    // support BF16 computations. Meanwhile, FP32 and FP16 elementwise
+    // instructions are not promoted and remain unchanged.
+    switch (dtype) {
+      case F16:
+        return F16;
+      case F32:
+      case BF16:
+      default:
+        return F32;
+    }
+  }
 
-  ENTRY convolution.test.bf16 {
-    p0 = bf16[8,4,5,5,1] parameter(0)
-    p1 = bf16[3,3,3,1,32] parameter(1)
-    ROOT conv = bf16[8,4,5,5,32] convolution(p0, p1), window={size=3x3x3 pad=1_1x1_1x1_1}, dim_labels=b012f_012io->b012f
-})";
+  void AdjustToleranceForDtype(PrimitiveType for_type, float atol_arg,
+                               float rtol_arg) {
+    if (dtype == for_type) {
+      atol = atol_arg;
+      rtol = rtol_arg;
+    }
+  }
 
-  EXPECT_TRUE(RunAndCompare(convolution_module_str, ErrorSpec{1e-4, 1e-4}));
-  MatchOptimizedHlo(convolution_module_str, conv_rewrite_str_);
-}
+  std::string PromotedDtypeToString() {
+    auto promoted_type = PromotedDtype();
+    return primitive_util::LowercasePrimitiveTypeName(promoted_type);
```

Comment:
Nit: We can merge both lines since we don't use `promoted_type` anywhere else.
```suggestion
    return primitive_util::LowercasePrimitiveTypeName(PromotedDtype());
```

Reply:
Done